clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetchRequestManagerTest.java

```
@@ -0,0 +1,3573 @@
+    }
+
+    @Test
+    public void testFetchOffsetOutOfRange() {
+        buildFetcher();
+        assignFromUser(singleton(tp0));
+        subscriptions.seek(tp0, 0);
+
+        assertEquals(1, sendFetches());
+        client.prepareResponse(fullFetchResponse(tidp0, records, Errors.OFFSET_OUT_OF_RANGE, 100L, 0));
+        networkClientDelegate.poll(time.timer(0));
+        assertEmptyFetch("Should not return records or advance position on fetch error");
+        assertTrue(subscriptions.isOffsetResetNeeded(tp0));
+        assertNull(subscriptions.validPosition(tp0));
+        assertNull(subscriptions.position(tp0));
+    }
+
+    @Test
+    public void testStaleOutOfRangeError() {
+        // verify that an out of range error which arrives after a seek
+        // does not cause us to reset our position or throw an exception
+        buildFetcher();
+        assignFromUser(singleton(tp0));
+        subscriptions.seek(tp0, 0);
+
+        assertEquals(1, sendFetches());
+        client.prepareResponse(fullFetchResponse(tidp0, records, Errors.OFFSET_OUT_OF_RANGE, 100L, 0));
+        subscriptions.seek(tp0, 1);
+        networkClientDelegate.poll(time.timer(0));
+        assertEmptyFetch("Should not return records or advance position on fetch error");
+        assertFalse(subscriptions.isOffsetResetNeeded(tp0));
+        assertEquals(1, subscriptions.position(tp0).offset);
+    }
+
+    @Test
+    public void testFetchedRecordsAfterSeek() {
+        buildFetcher(OffsetResetStrategy.NONE, new ByteArrayDeserializer(),
+                new ByteArrayDeserializer(), 2, IsolationLevel.READ_UNCOMMITTED);
+
+        assignFromUser(singleton(tp0));
+        subscriptions.seek(tp0, 0);
+
+        assertTrue(sendFetches() > 0);
+        client.prepareResponse(fullFetchResponse(tidp0, records, Errors.OFFSET_OUT_OF_RANGE, 100L, 0));
+        networkClientDelegate.poll(time.timer(0));
+        assertFalse(subscriptions.isOffsetResetNeeded(tp0));
```

It would be useful to add a comment why offset is not reset yet on `OFFSET_OUT_OF_RANGE`. This is because the error handling happens when new records are polled.

---

Done. Here's what I added:

```java
        // The partition is not marked as needing its offset reset because that error handling logic is
        // performed during the fetch collection. When we call seek() before we collect the fetch, the
        // partition's position is updated (to offset 2) which is different from the offset from which
        // we fetched the data (from offset 0).
```

I'll admit, I don't know that I totally understand what I wrote or if it's correct :)

---

Thanks for updating the comment, Kirk. The comment seems correct. It would be useful to further add that because the fetch position in the request is different from the one in fetch state, the fetched data, including the `OFFSET_OUT_OF_RANGE` error is ignored in `FetchCollector.handleInitializeErrors` because of the following code. 

`                if (position == null || fetchOffset != position.offset) {
`

It's a bit weird that the above check is only done for the `OFFSET_OUT_OF_RANGE` error, instead of any error. It might be useful to file a jira to revisit this logic.

---

I tried to update the comment, but ended up confused on how to explain it. Because this looks like an existing issue that needs more investigation, I filed KAFKA-15641 to follow up.